lib/Service/AntiAbuseService.php

```diff
@@ -0,0 +1,153 @@
+	private $cacheFactory;
+
+	/** @var ITimeFactory */
+	private $timeFactory;
+
+	/** @var LoggerInterface */
+	private $logger;
+
+	public function __construct(IConfig $config,
+								ICacheFactory $cacheFactory,
+								ITimeFactory $timeFactory,
+								LoggerInterface $logger) {
+		$this->config = $config;
+		$this->cacheFactory = $cacheFactory;
+		$this->timeFactory = $timeFactory;
+		$this->logger = $logger;
+	}
+
+	public function onBeforeMessageSent(IUser $user,
+										NewMessageData $messageData): void {
+		$abuseDetection = $this->config->getAppValue(
+			Application::APP_ID,
+			'abuse_detection',
+			'off'
+		);
+		if ($abuseDetection !== 'on') {
+			$this->logger->debug('Anti abuse detection is off');
+			return;
+		}
+
+		$this->checkNumberOfRecipients($user, $messageData);
+		$this->checkRateLimits($user, $messageData);
+	}
+
+	private function checkNumberOfRecipients(IUser $user,
+											 NewMessageData $messageData): void {
+		$numberOfRecipientsThreshold = (int)$this->config->getAppValue(
+			Application::APP_ID,
+			'abuse_number_of_recipients_per_message_threshold',
+			'0',
+		);
+		if ($numberOfRecipientsThreshold <= 1) {
+			return;
+		}
+
+		$actualNumberOfRecipients = count($messageData->getTo())
+			+ count($messageData->getCc())
+			+ count($messageData->getBcc());
+
+		if ($actualNumberOfRecipients >= $numberOfRecipientsThreshold) {
+			$this->logger->alert('User {user} sends to a suspicious number of recipients. {expected} are allowed. {actual} are used', [
+				'user' => $user->getUID(),
+				'expected' => $numberOfRecipientsThreshold,
+				'actual' => $actualNumberOfRecipients,
+			]);
+		}
+	}
+
+	private function checkRateLimits(IUser $user,
+									 NewMessageData $messageData): void {
+		if (!$this->cacheFactory->isAvailable()) {
+			// No cache, no rate limits
+			return;
+		}
+		$cache = $this->cacheFactory->createDistributed('mail_anti_abuse');
+		if (!($cache instanceof IMemcache)) {
+			// This integration only works with caches that support inc and dec
+			return;
+		}
+
+		$this->checkRateLimitsForPeriod($user, $messageData, $cache, '15m', 15 * 60);
+		$this->checkRateLimitsForPeriod($user, $messageData, $cache, '1h', 60 * 60);
+		$this->checkRateLimitsForPeriod($user, $messageData, $cache, '1d', 24 * 60 * 60);
+	}
+
+	private function checkRateLimitsForPeriod(IUser $user,
+											  NewMessageData $messageData,
+											  IMemcache $cache,
+											  string $id,
+											  int $period): void {
+		$maxNumberOfMessages = (int)$this->config->getAppValue(
+			Application::APP_ID,
+			'abuse_number_of_messages_per_' . $id,
+			'0',
+		);
+		if ($maxNumberOfMessages === 0) {
+			// No limit set
+			return;
+		}
+
+		$now = $this->timeFactory->getTime();
+
+		// Build blocks of periods per period size
+		$periodStart = ((int)($now / $period)) * $period;
+		$cacheKey = implode('_', ['counter', $id, $periodStart]);
+		$cache->add($cacheKey, 0);
+		$counter = $cache->inc($cacheKey);
```

Comment:
The increment should possibly be the number of recipients. Because if you send one message to five people then you cause five messages :thinking: